In [1]:
import os
import numpy as np
import pandas as pd

In [2]:
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set()

In [3]:
df = pd.read_csv(r'C:\Users\lilie\OneDrive\Bureau\ironHack\dataset\fake_real_final.csv')

In [4]:
df['txt'] = df['title'] + " " + df['text']
df = df.drop(columns = ['title', 'text','subject','date'])
df = df.loc[~df['txt'].isna()]
df.head()

label_REAL                                                txt
0           0  You Can Smell Hillary’s Fear Daniel Greenfield...
1           0  Watch The Exact Moment Paul Ryan Committed Pol...
2           1  Kerry to go to Paris in gesture of sympathy U....
3           0  Bernie supporters on Twitter erupt in anger ag...
4           1  The Battle of New York: Why This Primary Matte...

In [5]:
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize, RegexpTokenizer

# Lemmatization & stemming - on the DF

In [ ]:
#removing stop_words
#stops = stopwords.words('english')
#df['txt_without_stopwords'] = df['txt'].apply(lambda x: ' '.join([word for word in x.lower().split() if word not in (set(stopwords.words('english')))]))

In [12]:
df['txt_lower'] = df['txt'].apply(lambda x: x.lower())
df.head()

label_REAL                                                txt  \
0           0  You Can Smell Hillary’s Fear Daniel Greenfield...   
1           0  Watch The Exact Moment Paul Ryan Committed Pol...   
2           1  Kerry to go to Paris in gesture of sympathy U....   
3           0  Bernie supporters on Twitter erupt in anger ag...   
4           1  The Battle of New York: Why This Primary Matte...   

                                           txt_lower  
0  you can smell hillary’s fear daniel greenfield...  
1  watch the exact moment paul ryan committed pol...  
2  kerry to go to paris in gesture of sympathy u....  
3  bernie supporters on twitter erupt in anger ag...  
4  the battle of new york: why this primary matte...

In [13]:
#tokenization without punctuation
df['tokenised_txt'] = df['txt_lower'].apply(lambda x: RegexpTokenizer(r'\w+').tokenize(x))

In [14]:
#removing stop words
stops = stopwords.words('english')
df['txt_without_stopwords'] = df['tokenised_txt'].apply(lambda x: [w for w in x if not w.isdigit() and not w in stops])

label_REAL                                                txt  \
0           0  You Can Smell Hillary’s Fear Daniel Greenfield...   
1           0  Watch The Exact Moment Paul Ryan Committed Pol...   
2           1  Kerry to go to Paris in gesture of sympathy U....   
3           0  Bernie supporters on Twitter erupt in anger ag...   
4           1  The Battle of New York: Why This Primary Matte...   

                                           txt_lower  \
0  you can smell hillary’s fear daniel greenfield...   
1  watch the exact moment paul ryan committed pol...   
2  kerry to go to paris in gesture of sympathy u....   
3  bernie supporters on twitter erupt in anger ag...   
4  the battle of new york: why this primary matte...   

                                       tokenised_txt  \
0  [you, can, smell, hillary, s, fear, daniel, gr...   
1  [watch, the, exact, moment, paul, ryan, commit...   
2  [kerry, to, go, to, paris, in, gesture, of, sy...   
3  [bernie, supporters, on, twitter, erupt, in, a...   
4  [the, battle, of, new, york, why, this, primar...   

                               txt_without_stopwords  
0  [smell, hillary, fear, daniel, greenfield, shi...  
1  [watch, exact, moment, paul, ryan, committed, ...  
2  [kerry, go, paris, gesture, sympathy, u, secre...  
3  [bernie, supporters, twitter, erupt, anger, dn...  
4  [battle, new, york, primary, matters, primary,...

In [31]:
#lemmatization
lm = WordNetLemmatizer()
df['lemmatized_txt'] = df['txt_without_stopwords'].apply(lambda x:[lm.lemmatize(word) for word in x])

label_REAL                                                txt  \
0           0  You Can Smell Hillary’s Fear Daniel Greenfield...   
1           0  Watch The Exact Moment Paul Ryan Committed Pol...   
2           1  Kerry to go to Paris in gesture of sympathy U....   
3           0  Bernie supporters on Twitter erupt in anger ag...   
4           1  The Battle of New York: Why This Primary Matte...   

                                           txt_lower  \
0  you can smell hillary’s fear daniel greenfield...   
1  watch the exact moment paul ryan committed pol...   
2  kerry to go to paris in gesture of sympathy u....   
3  bernie supporters on twitter erupt in anger ag...   
4  the battle of new york: why this primary matte...   

                                       tokenised_txt  \
0  [you, can, smell, hillary, s, fear, daniel, gr...   
1  [watch, the, exact, moment, paul, ryan, commit...   
2  [kerry, to, go, to, paris, in, gesture, of, sy...   
3  [bernie, supporters, on, twitter, erupt, in, a...   
4  [the, battle, of, new, york, why, this, primar...   

                               txt_without_stopwords  \
0  [smell, hillary, fear, daniel, greenfield, shi...   
1  [watch, exact, moment, paul, ryan, committed, ...   
2  [kerry, go, paris, gesture, sympathy, u, secre...   
3  [bernie, supporters, twitter, erupt, anger, dn...   
4  [battle, new, york, primary, matters, primary,...   

                                      lemmatized_txt  
0  [smell, hillary, fear, daniel, greenfield, shi...  
1  [watch, exact, moment, paul, ryan, committed, ...  
2  [kerry, go, paris, gesture, sympathy, u, secre...  
3  [bernie, supporter, twitter, erupt, anger, dnc...  
4  [battle, new, york, primary, matter, primary, ...

In [16]:
#stemming
from nltk.stem import PorterStemmer
ps = PorterStemmer()
df['stem_txt'] = df['txt_without_stopwords'].apply(lambda x:[ps.stem(word) for word in x])
df.head()

label_REAL                                                txt  \
0           0  You Can Smell Hillary’s Fear Daniel Greenfield...   
1           0  Watch The Exact Moment Paul Ryan Committed Pol...   
2           1  Kerry to go to Paris in gesture of sympathy U....   
3           0  Bernie supporters on Twitter erupt in anger ag...   
4           1  The Battle of New York: Why This Primary Matte...   

                                           txt_lower  \
0  you can smell hillary’s fear daniel greenfield...   
1  watch the exact moment paul ryan committed pol...   
2  kerry to go to paris in gesture of sympathy u....   
3  bernie supporters on twitter erupt in anger ag...   
4  the battle of new york: why this primary matte...   

                                       tokenised_txt  \
0  [you, can, smell, hillary, s, fear, daniel, gr...   
1  [watch, the, exact, moment, paul, ryan, commit...   
2  [kerry, to, go, to, paris, in, gesture, of, sy...   
3  [bernie, supporters, on, twitter, erupt, in, a...   
4  [the, battle, of, new, york, why, this, primar...   

                               txt_without_stopwords  \
0  [smell, hillary, fear, daniel, greenfield, shi...   
1  [watch, exact, moment, paul, ryan, committed, ...   
2  [kerry, go, paris, gesture, sympathy, u, secre...   
3  [bernie, supporters, twitter, erupt, anger, dn...   
4  [battle, new, york, primary, matters, primary,...   

                                            stem_txt  
0  [smell, hillari, fear, daniel, greenfield, shi...  
1  [watch, exact, moment, paul, ryan, commit, pol...  
2  [kerri, go, pari, gestur, sympathi, u, secreta...  
3  [berni, support, twitter, erupt, anger, dnc, t...  
4  [battl, new, york, primari, matter, primari, d...

In [17]:
df_filtered = df.drop(['txt','txt_lower','tokenised_txt'], axis=1).copy()

In [88]:
print(type(df_filtered.loc[0, 'lemmatized_txt']))

<class 'list'>


In [19]:
#df_filtered['lemmatized_txt'] = df_filtered['lemmatized_txt'].str.join(',')
df_filtered['stem_txt'] = df_filtered['stem_txt'].str.join(',')

In [22]:
df_filtered.head()

label_REAL                                           stem_txt
0           0  smell,hillari,fear,daniel,greenfield,shillman,...
1           0  watch,exact,moment,paul,ryan,commit,polit,suic...
2           1  kerri,go,pari,gestur,sympathi,u,secretari,stat...
3           0  berni,support,twitter,erupt,anger,dnc,tri,warn...
4           1  battl,new,york,primari,matter,primari,day,new,...

In [23]:
#df_filtered.to_csv(r'C:\Users\lilie\GitHubProjects\IronEstelle\FinalProject\datasets\df_lemmatized.csv', index = False)
df_filtered.to_csv(r'C:\Users\lilie\GitHubProjects\IronEstelle\FinalProject\datasets\df_stemmed.csv', index = False)

# Split train & test

In [24]:
#X = df_filtered.lemmatized_txt
X = df_filtered.stem_txt
y= df_filtered.label_REAL

In [25]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.3, random_state=0)

# Tf-idf on lemmatized text

In [94]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [95]:
### tf-idf
tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_df=0.7)

X_tfidf_train = tfidf_vectorizer.fit_transform(X_train) 
X_tfidf_test = tfidf_vectorizer.transform(X_test)

In [121]:
tfidf_vectorizer2 = TfidfVectorizer(stop_words='english', max_df=0.7)

X_tfidf_train2 = tfidf_vectorizer2.fit_transform(X_train) 
X_tfidf_test2 = tfidf_vectorizer.transform(X_test)

# Modelling

In [102]:
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import MultinomialNB

In [103]:
#passive aggressive 
clf = PassiveAggressiveClassifier(n_iter_no_change=50)
clf.fit(X_tfidf_train, y_train)
y_pred = clf.predict(X_tfidf_test)

display(cm)
print('Accuracy is', accuracy_score(y_test, y_pred).round(4))
print('Precision is', precision_score(y_test, y_pred).round(4))
print('Recall is', recall_score(y_test, y_pred).round(4))
print('F1 is', f1_score(y_test, y_pred).round(4))

array([[7796,  192],
       [ 236, 7075]], dtype=int64)

Accuracy is 0.9729
Precision is 0.9749
Recall is 0.9681
F1 is 0.9715


In [104]:
#decision tree
dtc = DecisionTreeClassifier(criterion='entropy')
dtc.fit(X_tfidf_train, y_train)
y_pred=dtc.predict(X_tfidf_test)

display(cm)
print('Accuracy is', accuracy_score(y_test, y_pred).round(4))
print('Precision is', precision_score(y_test, y_pred).round(4))
print('Recall is', recall_score(y_test, y_pred).round(4))
print('F1 is', f1_score(y_test, y_pred).round(4))

array([[7796,  192],
       [ 236, 7075]], dtype=int64)

Accuracy is 0.9497
Precision is 0.9533
Recall is 0.9409
F1 is 0.9471


In [ ]:
# naive bayes
classifier=MultinomialNB()
classifier.fit(X_tfidf_train, y_train)
y_pred2 = classifier.predict(X_tfidf_test)

display(cm)
print('Accuracy is', accuracy_score(y_test, y_pred).round(4))
print('Precision is', precision_score(y_test, y_pred).round(4))
print('Recall is', recall_score(y_test, y_pred).round(4))
print('F1 is', f1_score(y_test, y_pred).round(4))

# Tf-idf on stemmed text

In [26]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer(strip_accents=ascii, stop_words='english', max_df=0.9)
X_tfidf_train = tfidf_vectorizer.fit_transform(X_train) 
X_tfidf_test = tfidf_vectorizer.transform(X_test)

In [27]:
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score

clf = PassiveAggressiveClassifier(n_iter_no_change=100)
clf.fit(X_tfidf_train, y_train)
y_pred = clf.predict(X_tfidf_test)

cm = confusion_matrix(y_test, y_pred)
display(cm)
print('Accuracy is', accuracy_score(y_test, y_pred).round(4))
print('Precision is', precision_score(y_test, y_pred).round(4))
print('Recall is', recall_score(y_test, y_pred).round(4))
print('F1 is', f1_score(y_test, y_pred).round(4))

array([[7779,  209],
       [ 240, 7071]], dtype=int64)

Accuracy is 0.9707
Precision is 0.9713
Recall is 0.9672
F1 is 0.9692


# Conclusion

With passive aggressive classifier:
   *  stemming + tfidf vectorizer : precision = 0.9713
   *  lemmatization + tfidf vectorizer : precision = 0.9749
   *  **tfidf vectorizer : precision > 0.9755**
   
Applying tfidf only is enought to get good precision.